src/torchmetrics/detection/iou.py

```diff
@@ -0,0 +1,185 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+from torch import Tensor
+
+from torchmetrics.detection.helpers import _fix_empty_tensors, _input_validator
+from torchmetrics.functional.detection.iou import _iou_compute, _iou_update
+from torchmetrics.metric import Metric
+from torchmetrics.utilities.data import dim_zero_cat, dim_zero_mean
+
+# from torchmetrics.utilities import rank_zero_warn
+from torchmetrics.utilities.imports import _TORCHVISION_GREATER_EQUAL_0_8
+
+if _TORCHVISION_GREATER_EQUAL_0_8:
+    from torchvision.ops import box_convert
+else:
+    box_convert = None
+
+
+class IntersectionOverUnion(Metric):
+    r"""Computes Intersection Over Union (IoU).
+
+    Args:
+        box_format:
+            Input format of given boxes. Supported formats are ``[`xyxy`, `xywh`, `cxcywh`]``.
+        iou_thresholds:
+            Optional IoU thresholds for evaluation. If set to `None` the threshold is ignored.
+        class_metrics:
+            Option to enable per-class metrics for IoU. Has a performance impact.
+        respect_labels:
+            Replace IoU values with the `invalid_val` if the labels do not match.
+        kwargs:
+             Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+    """
+    is_differentiable: bool = False
+    higher_is_better: Optional[bool] = True
+    full_state_update: bool = True
+
+    detections: List[Tensor]
+    detection_scores: List[Tensor]
+    detection_labels: List[Tensor]
+    groundtruths: List[Tensor]
+    groundtruth_labels: List[Tensor]
+    results: List[Tensor]
+    labels_eq: List[bool]
+
+    iou_update_fn: Callable[[Tensor, Tensor, bool, float], Tensor] = _iou_update
+    iou_compute_fn: Callable[[Tensor, bool], Tensor] = _iou_compute
+    iou_type: str = "iou"
+    invalid_val: float = 0
```

Comment:
This is a nice patter event it may be a danger for runtime changes... :)
cc: @SkafteNicki